### PR TITLE
net: shell: Allow to set interface for ping (add -I option)

### PR DESCRIPTION
**Pull request by lmajewski**
This change adds new switch to 'net ping' command - to be more specific
the '-I <iface>', where the <iface> is the number of supported network
interface.

One can discover this number from the Zephyr shell after running
'net iface'.

The default behavior - i.e. without the -I switch present is preserved.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

### CODE REVIEW

File: subsys/net/ip/net_shell.c [hunk 1]
```diff
@@ -2933,6 +2934,9 @@ static int ping_ipv6(const struct shell *shell,
 #if defined(CONFIG_NET_ROUTE)
 	struct net_route_entry *route;
 #endif
+	if (iface_idx != -1) {
+		iface = net_if_get_by_index(iface_idx);
+	}
```
Comment (jukkar) [inline subsys/net/ip/net_shell.c:2939]: This can be removed.

And then below, we can do things like this. So if the iface is not set, we figure it out from the destination address and if everything fails, only then use the defaut interface
```
if (!iface) {
    iface = net_if_ipv6_select_src_iface(...);
    if (!iface) {
    ...
         if (nbr) {
          ...
         } else {
            iface = net_if_get_default();
         }
    }
}
```

File: subsys/net/ip/net_shell.c [hunk 2]
```diff
@@ -3047,16 +3051,22 @@ static enum net_verdict handle_ipv4_echo_reply(struct net_pkt *pkt,
 static int ping_ipv4(const struct shell *shell,
 		     char *host,
 		     unsigned int count,
-		     unsigned int interval)
+		     unsigned int interval,
+		     int iface_idx)
 {
 	struct in_addr ipv4_target;
+	struct net_if *iface;
```
Comment (jukkar) [inline subsys/net/ip/net_shell.c:3058]: Similar changes to ipv4, so here `struct net_if *iface = net_if_get_by_index(iface_idx);`

File: subsys/net/ip/net_shell.c [hunk 3]
```diff
@@ -3047,16 +3051,22 @@ static enum net_verdict handle_ipv4_echo_reply(struct net_pkt *pkt,
...
 	if (net_addr_pton(AF_INET, host, &ipv4_target) < 0) {
 		return -EINVAL;
 	}
 
-	struct net_if *iface = net_if_ipv4_select_src_iface(&ipv4_target);
+	if (iface_idx == -1) {
+		iface = net_if_ipv4_select_src_iface(&ipv4_target);
+	} else {
+		iface = net_if_get_by_index(iface_idx);
+	}
```
Comment (jukkar) [inline subsys/net/ip/net_shell.c:3069]: and here
```
if (!iface) {
     iface = net_if_ipv4_select_src_iface(&ipv4_target);
}
```

File: subsys/net/ip/net_shell.c [hunk 4]
```diff
@@ -4470,7 +4489,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(net_cmd_vlan,
 
 SHELL_STATIC_SUBCMD_SET_CREATE(net_cmd_ping,
 	SHELL_CMD(--help, NULL,
-		  "'net ping [-c count] [-i interval ms] <host>' "
+		  "'net ping [-c count] [-i interval ms] [-I <iface>] <host>' "
```
Comment (jukkar) [inline subsys/net/ip/net_shell.c:4492]: change this to `[-I <iface index>]` so it is clear what the value should be

**Review by jukkar — CHANGES_REQUESTED** [review]
Looks good, minor changes needed.

**Comment by pfalcon** [comment]
May I suggest to remove word "default" from the commit title, i.e. make it something like:

> net: shell: Allow to set interface for ping (add -I <iface> option) 

Saying that it sets default interface for ping is confusing, indeed, commit the message says further:

> The default behavior - i.e. without the -I switch present is preserved.

I.e. the default interface stays the same, it's just as switch is added too override the used interface for a particular "ping" invocation.

**Comment by pfalcon** [comment]
> net: shell: Allow to set interface for ping (add -I option)

Thanks.

File: subsys/net/ip/net_shell.c [hunk 5]
```diff
@@ -2923,28 +2923,32 @@ static enum net_verdict handle_ipv6_echo_reply(struct net_pkt *pkt,
...
-	struct net_if *iface = net_if_get_default();
+	struct net_if *iface = net_if_get_by_index(iface_idx);
 	int ret = 0;
 	struct in6_addr ipv6_target;
 	struct net_nbr *nbr;
 
 #if defined(CONFIG_NET_ROUTE)
 	struct net_route_entry *route;
 #endif
-
```
Comment (jukkar) [inline subsys/net/ip/net_shell.c:2936]: Unrelated change, please do not remove the line here.

**Review by jukkar — CHANGES_REQUESTED** [review]
Thanks!

**Review by jukkar — APPROVED** [review]

**Review by pfalcon — APPROVED** [review]
